readme.md
# CLEO Redux

[![Discord](https://img.shields.io/discord/911487285990674473?style=for-the-badge)](https://discord.gg/d5dZSfgBZr)
[![YouTube Channel](https://img.shields.io/badge/YouTube-Channel-FF0000?style=for-the-badge)](https://www.youtube.com/playlist?list=PLNxQuEFtVkeizoLEQiok7qzr1f0mcwfFb)

- [Начало](#начало)
	- [Что такое CLEO Redux?](#что-такое-cleo-redux)
	- [Поддерживаемые языки](#поддерживаемые-языки)
	- [Поддерживаемые релизы](#поддерживаемые-релизы)
	- [Связь с CLEO Library](#связь-с-cleo-library)
		- [Запуск CLEO Redux как отдельного ПО](#запуск-cleo-redux-как-отдельного-по)
		- [Запуск CLEO Redux в качестве дополнения к библиотеке CLEO](#запуск-cleo-redux-в-качестве-дополнения-к-библиотеке-cleo)
- [Установка](#установка)
	- [Первичная настройка](#первичная-настройка)
	- [Совместимость с re3 и reVC](#совместимость-с-re3-и-revc)
	- [Совместимость с The Trilogy: The Definitive Edition](#совместимость-с-the-trilogy-the-definitive-edition)
	- [Удаление](#удаление)
- [Конфигурация](#конфигурация)
	- [Общая конфигурация](#общая-конфигурация)
	- [Разрешения](#разрешения)
		- [Все](#все)
		- [Слабые](#слабые)
		- [Строгие](#строгие)
		- [Исключения](#исключения)

- [Лог](#лог)
- [Пользовательские сценарии](#пользовательские-сценарии)
	- [Добавление нового скрипта](#добавление-нового-скрипта)
	- [Удаление скрипта](#удаление-скрипта)
	- [Пользовательские команды](#пользовательские-команды)
	- [Написание CS-скриптов](#написание-cs-скриптов)
	- [Написание JS-скриптов](#написание-js-скриптов)
	- [Интеграция с Visual Studio Code](#интеграция-с-visual-studio-code)
	- [Интеграция с Atom](#интеграция-с-atom)
- [Поддержка JavaScript](#поддержка-javascript)
	- [Предварительное условие](#предварительное-условие)
	- [Жизненный цикл скрипта](#жизненный-цикл-скрипта)
	- [Собственные команды](#собственные-команды)
		- [Класс ScriptObject против объекта](#класс-scriptobject-против-объекта)
		- [Мат. классы](#мат-классы)
		- [JSON](#json)    
		- [Свободный интерфейс](#свободный-интерфейс)
		- [Примеры](#примеры)
	- [Импорт](#импорт)
	- [Пользовательские привязки](#пользовательские-привязки)
	- [Устаревшие](#устаревшие)
- [Пользовательский текст](#пользовательский-текст)
- [Особенности разработки](#особенности-разработки)
	- [Лог SCM](#лог-scm)
	- [Горячая перезагрузка](#горячая-перезагрузка)
	- [SDK](#sdk)
- [Лицензия](#лицензия)

## Начало

### Что такое CLEO Redux?

CLEO Redux — это среда выполнения сценариев для игр эпохи GTA 3D. Это полноценный член семейства CLEO, который предоставляет знакомый опыт всем, кто использовал библиотеку CLEO для классической GTA San Andreas или ее повторные реализации для других игр. Основная цель CLEO — предоставить возможность легко настраивать игру с помощью многочисленных пользовательских скриптов.

Если вы новичок в CLEO, посетите [официальный сайт](https://cleo.li/), чтобы найти больше информации об этом.

### Поддерживаемые языки

CLEO Redux поддерживает скомпилированные бинарные скрипты (`*.cs`) в родном формате SCM и простые текстовые скрипты (`*.js`), написанные на JavaScript.

CLEO Redux ориентируется на JavaScript как на основной язык для пользовательских скриптов. JavaScript — популярный язык программирования с богатой экосистемой и большим количеством доступной информации. Он свободен от ограничений и ловушек языка SCM, таких как отсутствие поддержки функций, массивов или малое количество переменных.

### Поддерживаемые релизы

Классические:

- GTA III 1.0
- GTA Vice City 1.0
- GTA San Andreas 1.0 (только с [CLEO 4.4](https://github.com/cleolibrary/CLEO4))

Ремастеры (The Trilogy):

- GTA III 1.0.0.14718
- GTA Vice City 1.0.0.14718
- San Andreas 1.0.0.14296, 1.0.0.14388, 1.0.0.14718 (Название обновления 1.03)

Подробнее см. [Часто задаваемые вопросы по окончательному изданию](https://github.com/TheFantomKiller420/CLEO-Redux/blob/master/the-definitive-edition-faq.md).

Остальные:

- re3 (см. [Сведения о совместимости](#совместимость-с-re3-и-revc))
- reVC (см. [Сведения о совместимости](#совместимость-с-re3-и-revc))

CLEO Redux поддерживает только ПК-версию каждой игры.

Для получения полной информации о поддерживаемых функциях [обратитесь к этой странице](https://github.com/cleolibrary/CLEO-Redux/wiki/Feature-Support-Matrix). Также существуют известные ограничения [перечислены здесь](unsupported.md).

### Связь с CLEO Library

CLEO — это общее название пользовательских библиотек, разработанных и созданных для GTA III, Vice City или San Andreas. Каждую версию можно найти и скачать [здесь](https://cleo.li/download.html). CLEO Redux — это _другая_ реализация CLEO, созданная с нуля, с несколькими отличительными особенностями, такими как единая кодовая база для всех игр или поддержка кода JavaScript.

На данный момент CLEO Redux не может рассматриваться как полноценная замена CLEO Library из-за отсутствия поддержки многих широко используемых CLEO-команд. Чтобы решить эту проблему и получить максимальную отдачу от двух библиотек, CLEO Redux поддерживает две разные стратегии использования.

CLEO Redux может работать как отдельное ПО или как дополнение к CLEO Library. В первом случае ваша директория с игрой будет содержать только файл `cleo_redux.asi` (или `cleo_redux64.asi`). Во втором случае в вашем каталоге с игрой будут и `cleo.asi` (или `III.CLEO.asi`, и `VC.CLEO.asi`), и `cleo_redux.asi` (или `cleo_redux64.asi`).

#### Запуск CLEO Redux как отдельного ПО

Как отдельное ПО CLEO Redux запускает скомпилированные скрипты и JavaScript и предоставляет доступ ко всем командам собственного скрипта. Он также предоставляет ограниченный набор [настраиваемых команд](#пользовательские-команды), обратно совместимых с библиотекой CLEO. Существующие CLEO-скрипты могут не работать, если они используют пользовательские команды (например, из стороннего плагина), не поддерживаемые CLEO Redux.

Этот режим не работает в классической GTA San Andreas.

#### Запуск CLEO Redux в качестве дополнения к библиотеке CLEO

В качестве дополнения CLEO Redux работает вместе с CLEO Library, делегируя ему всю заботу о пользовательских скриптах. Это означает, что все пользовательские скрипты и плагины, созданные для библиотеки CLEO, будут продолжать работать точно так же. CLEO Redux управляет только JS-скриптами. Все пользовательские команды становятся доступными для среды выполнения JavaScript, что означает, что вы можете использовать команды, которые в настоящее время не реализованы нативно в CLEO Redux, например, для [файлов](https://library.sannybuilder.com/#/gta3/classes/File) или [ динамические библиотеки](https://library.sannybuilder.com/#/gta3/classes/DynamicLibrary).

Этот режим работает в классической GTA III, GTA Vice City и GTA San Andreas, где существует библиотека CLEO.

## Установка

Если вы используете The Definitive Edition (GTA III, VC или SA):

- [Следуйте этому руководству](the-definitive-edition-faq.md#how-to-install-cleo-redux-in-the-definition-edition).

Для всех остальных игр:

- Скопируйте `cleo_redux.asi` в папку с игрой.

- Запустите игру

Примечание: CLEO Redux не изменяет игровые файлы. Он использует тот факт, что игра изначально загружает файлы `.asi` в качестве дополнений к библиотеке Miles Sound System. Никакого дополнительного ПО не требуется.

### Первичная настройка

Во время первого запуска игры может быть заметное отставание, так как CLEO Redux загружает файлы, необходимые для [поддержки JavaScript](#поддержка-javascript). При последующих запусках этого не произойдет.

Также в папке с игрой\* должна появиться новая папка с именем `CLEO`. Это основное место для всех CLEO-скриптов, плагинов и конфигов.

\*Если CLEO не может создать новые файлы в каталоге игры из-за отсутствия прав на запись, он использует альтернативный путь в `C:\Users\<ваше_имя_пользователя>\AppData\Roaming\CLEO Redux`. Там можно найти `cleo_redux.log` и каталог `CLEO`.

### Совместимость с re3 и reVC

CLEO Redux поддерживает только «Windows D3D9 MSS 32bit» версию `re3` или `reVC`.

При запуске на `re3` и `reVC` убедитесь, что в каталоге игры есть файл `re3.pdb` (для **re3**) или `reVC.pdb` (для **reVC**). Из-за динамического характера адресов памяти в этих реализациях CLEO Redux использует отладочную информацию, хранящуюся в файле PDB, для правильного определения своего местоположения.

### Совместимость с The Trilogy: The Definitive Edition

Проверьте [Часто задаваемые вопросы по окончательному изданию](the-definitive-edition-faq.md)

### Удаление

- Удалите `cleo_redux.asi` (или `cleo_redux64.asi`).
- Удалите папку `CLEO` (необязательно).
- Удалите `cleo_redux.log` (необязательно).

## Конфигурация

CLEO Redux предоставляет некоторые настраиваемые параметры в файле `CLEO\.config\cleo.ini`.

### Общая конфигурация

- `AllowCs` - при значении `1` CLEO загружает и выполняет файлы `*.cs`, расположенные в каталоге CLEO. Включено по умолчанию.
- `AllowJs` - при значении `1` CLEO загружает и выполняет файлы `*.js`, расположенные в каталоге CLEO. Включено по умолчанию.
- `AllowFxt` - при значении `1` CLEO загружает и [использует](#пользовательский-текст) файлы `*.fxt`, расположенные в каталоге CLEO\CLEO_TEXT. Включено по умолчанию.
- `CheckUpdates` - при значении `1` CLEO проверяет наличие нового обновления, доступного для скачивания, при запуске игры. Включено по умолчанию.
- `LogOpcodes` - при значении `1` CLEO регистрирует все выполненные опкоды в пользовательских скриптах. Этот журнал является частью файла `cleo_redux.log` (см. [Лог](#лог))
- `PermissionLevel` - устанавливает уровень разрешений для небезопасных операций (см. ниже). По умолчанию используется «слабый».

### Разрешения

CLEO Redux признает некоторые [пользовательские команды](#пользовательские-команды) (коды операций) небезопасными и требует от пользователя решить, запускать их или нет. Необработанный доступ к памяти процесса, загрузка внешних библиотек или выполнение сетевых запросов могут быть вредными и вызывать нежелательные побочные эффекты. Поэтому CLEO вводит уровни разрешений для запуска небезопасного кода.

Доступны четыре уровня:

#### Все

Допускаются любые небезопасные операции. Используйте это только в том случае, если вы доверяете сценариям, которые запускаете.

#### Слабые

Это уровень разрешений по умолчанию.

Никакая небезопасная операция не разрешена, если сценарий явно не запрашивает ее. В настоящее время для запроса разрешения имя файла сценария должно включать маркеры разрешений, заключенные в квадратные скобки.

Например, если скрипт хочет получить доступ к памяти через `0A8D READ_MEMORY`, имя файла должно содержать `[mem]`, т.е. `крутой спаунер[mem].cs`. Если файл называется по-другому, CLEO отвергает `0A8D`, и скрипт вылетает.

#### Строгие

Никакая небезопасная операция не разрешена, если сценарий явно не запрашивает ее (см. `"Слабые"`), а файл конфигурации CLEO разрешает этот тип операции в разделе `Разрешения`. 

Раздел разрешений в `cleo.ini` позволяет включать или отключать группы небезопасных операций с помощью токенов разрешений. Например,

```ini
mem=0
```

отключает все коды операций, связанные с памятью, даже если скрипт имеет токен `[mem]` в имени файла.

Примечание. Раздел `Permissions` в `cleo.ini` вступает в силу только в том случае, если `PermissionLevel` имеет значение `Strict`.

#### Исключения

Небезопасная работа не допускается.

## Лог

CLEO регистрирует важные события и ошибки в файле `cleo_redux.log` расположенном в папке с игрой (или `C:\Users\<ваше_имя_пользователя>\AppData\Roaming\CLEO Redux`, см. [Первичная настройка](#первичная-настройка) примечание). Этот файл перезаписывается при каждом запуске игры. Если у вас возникнут какие-либо проблемы при использовании CLEO Redux, начните исследовать основную причину с этого файла.

Чтобы транслировать события в вашем терминале во время тестирования скрипта, запустите:

```
tail -f cleo_redux.log
```

`tail` - это команда unix, поэтому необходима совместимая среда (например, Git Bash).

## Пользовательские сценарии

### Добавление нового скрипта

Обычно файл скрипта нужно просто скопировать в каталог `CLEO`. Для установки некоторых скриптов могут потребоваться дополнительные действия. В случае каких-либо проблем проверьте документацию скрипта или обратитесь к его автору. 

### Удаление скрипта

Удалите файл скрипта из каталога `CLEO`. Некоторые сценарии могут потребовать дополнительных действий для отмены установки. В случае каких-либо проблем проверьте документацию скрипта или обратитесь к его автору. 

### Пользовательские команды

Примечание: Следующие команды предназначены только для классических игр. Для The Definitive Edition [проверьте эту информацию](the-definitive-edition-faq.md#can-i-use-cleo-opcodes).

- 0A8C [WRITE_MEMORY](https://library.sannybuilder.com/#/gta3/CLEO/0A8C) (**UNSAFE** - требует `mem` разрешение)
- 0A8D [READ_MEMORY](https://library.sannybuilder.com/#/gta3/CLEO/0A8D) (**UNSAFE** - требует `mem` разрешение)
- 0A8E [INT_ADD](https://library.sannybuilder.com/#/gta3/CLEO/0A8E)
- 0A8F [INT_SUB](https://library.sannybuilder.com/#/gta3/CLEO/0A8F)
- 0A90 [INT_MUL](https://library.sannybuilder.com/#/gta3/CLEO/0A90)
- 0A91 [INT_DIV](https://library.sannybuilder.com/#/gta3/CLEO/0A91)
- 0A93 [TERMINATE_THIS_CUSTOM_SCRIPT](https://library.sannybuilder.com/#/gta3/CLEO/0A93)
- 0AA5 [CALL_FUNCTION](https://library.sannybuilder.com/#/gta3/CLEO/0AA5) (**UNSAFE** - требует `mem` разрешение)
- 0AA6 [CALL_FUNCTION_RETURN](https://library.sannybuilder.com/#/gta3/CLEO/0AA6) (**UNSAFE** - требует `mem` разрешение)
- 0AA7 [CALL_METHOD](https://library.sannybuilder.com/#/gta3/CLEO/0AA7) (**UNSAFE** - требует `mem` разрешение)
- 0AA8 [CALL_METHOD_RETURN](https://library.sannybuilder.com/#/gta3/CLEO/0AA8) (**UNSAFE** - требует `mem` разрешение)
- 0AB0 [IS_KEY_PRESSED](https://library.sannybuilder.com/#/gta3/CLEO/0AB0)

### Написание CS-скриптов

Используйте [Sanny Builder 3](https://sannybuilder.com) в режимах редактирования GTA III, GTA VC или GTA SA соответственно. Посетите [эту страницу](https://cleo.li/scripts.html) для получения дополнительной информации.

[Проверьте Часто Задаваемые Вопросы](the-definitive-edition-faq.md#how-do-i-compile-cleo-scripts-with-sanny-builder) для получения информации о поддержке CS в обновленных играх.

### Написание JS-скриптов

Используйте VS Code (рекомендуется) или любой редактор по вашему выбору. Создайте новый файл с расширением `.js` и поместите его в папку CLEO. Дополнительную информацию см. в разделе [Поддержка JavaScript](#поддержка-javascript).

Среда выполнения поддерживает сценарии в [стандарте ECMAScript 2020](https://262.ecma-international.org/11.0/). Это означает, что вы можете использовать самые последние функции JavaScript из коробки, такие как импорт, классы, стрелочные функции и т.д.

CLEO Redux — это не Node.js. Не ожидайте, что здесь будут доступны сокеты, операции с файловой системой или другие функции Node.js.

### Интеграция с Visual Studio Code

Посмотреть демонстрацию: https://youtu.be/jqz8_lGnG4g

CLEO Redux генерирует типизации для всех поддерживаемых команд, которые вы можете использовать при написании JavaScript в VS Code. Добавьте следующую строку в свой скрипт `*.js` чтобы получить полную поддержку автозаполнения:

Для `GTA III` или `re3`: 

```
/// <reference path=".config/gta3.d.ts" />
```

Для `Vice City` или `reVC`

```
/// <reference path=".config/vc.d.ts" />
```

Для `San Andreas`

```
/// <reference path=".config/sa.d.ts" />
```

Эта строка указывает VS Code, где искать определения команд для функции автозаполнения. Путь может быть относительным относительно файла сценария или быть абсолютным. [Дополнительную информацию](https://www.typescriptlang.org/docs/handbook/triple-slash-directives.html#-reference-path-) на официальном портале TypeScript.

### Интеграция с Atom

Вся информация [тут](https://github.com/TheFantomKiller420/CLEO-Redux-integration-with-Atom).

Примечание: Советую пользоваться оригинальной [интеграцией с Visual Studio Code](#интеграция-с-visual-studio-code) ибо интеграция с Atom сделана любителем (мной), к тому же она находится на стадии разработки.

## Поддержка JavaScript

### Предварительное условие

Когда JavaScript включен, CLEO Redux требует определения команд с https://library.sannybuilder.com/. При первом запуске CLEO пытается загрузить их и поместить в вашу локальную директорию `CLEO/.config`. Если этого не произошло или вы не хотите, чтобы CLEO совершал сетевые вызовы, вручную скачайте необходимый файл (см. таблицу ниже) и поместите его в каталог `CLEO/.config`. 

| Игра                                     | Файл                                                                                                 | Минимальная требуемая версия |
| ---------------------------------------- | ---------------------------------------------------------------------------------------------------- | ---------------------------- |
| GTA III, re3                             | [gta3.json](https://github.com/sannybuilder/library/blob/master/gta3/gta3.json)                      | `0.218`                      |
| GTA VC, reVC                             | [vc.json](https://github.com/sannybuilder/library/blob/master/vc/vc.json)                            | `0.220`                      |
| GTA San Andreas (Классическая) 1.0       | [sa.json](https://github.com/sannybuilder/library/blob/master/sa/sa.json)                            | `0.237`                      |
| GTA III: The Definitive Edition          | [gta3_unreal.json](https://github.com/sannybuilder/library/blob/master/gta3_unreal/gta3_unreal.json) | `0.213`                      |
| Vice City: The Definitive Edition        | [vc_unreal.json](https://github.com/sannybuilder/library/blob/master/vc_unreal/vc_unreal.json)       | `0.215`                      |
| San Andreas: The Definitive Edition      | [sa_unreal.json](https://github.com/sannybuilder/library/blob/master/sa_unreal/sa_unreal.json)       | `0.221`                      |

### Жизненный цикл скрипта

Файл с кодом JavaScript должен иметь расширение `*.js` и содержать известные инструкции, как описано ниже. Скрипт может не иметь инструкций (пустой скрипт). Он запускается, как только начинается новая игра или загружается файл сохранения.

Скрипт завершается автоматически после выполнения последней инструкции. Среда выполнения также завершает зависшие сценарии, чтобы игра не зависала. Застрявший скрипт — это тот, которому потребовалось более 2 секунд для запуска с момента последней команды ожидания. Если это произошло, проверьте свои циклы, в некоторых из них отсутствует команда ожидания.

```js
while (true) {
  // бессмысленный бесконечный цикл, обычно зависающий в игре
  // будет завершено через две секунды
}
```

Среда выполнения завершит этот скрипт. Чтобы этого избежать, добавьте команду ожидания

```js
while (true) {
  wait(250);
  // все равно бессмысленно, но не зависает в игре
}
```

### Собственные команды

CLEO Redux поддерживает все внутриигровые команды (коды операций) в форме класса, как определено в библиотеке Sanny Builder.

#### Класс ScriptObject против объекта

Библиотека Sanny Builder определяет статический класс `Object` для группировки команд, позволяющих создавать и управлять трехмерными объектами в игре. В то же время в JavaScript есть [собственный класс Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object) со своими методами.

Чтобы не смешивать их, CLEO Redux использует класс `ScriptObject` вместо класса `Object` из библиотеки с [тем же интерфейсом](https://library.sannybuilder.com/#/gta3/classes/Object).

```js
var x = ScriptObject.Create(modelId, x, y, z); // код операции 0107, создает новый объект в игре

var x = Object.create(null); // собственный код JavaScript, создает новый объект в памяти JS
```

#### Мат. классы

Подобно классу `Object` (см. выше), и библиотека, и собственная среда выполнения JavaScript используют одно и то же имя для математических утилит: класс `Math`. Однако в данном случае было принято решение сохранить как собственные методы, так и скриптовые методы под одним и тем же именем класса.

Приоритет отдавался нативному коду в тех случаях, когда он обеспечивал ту же функциональность, что и опкоды скрипта. Например, чтобы вычислить абсолютное значение числа, используйте собственный [Math.abs](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/abs), а не [Math.Abs](https://library.sannybuilder.com/#/gta3?q=%22abs%22). См. [Использование математики](using-math.md) для более подробной информации.

#### JSON

В [импорте](#импорт) сказано что импортировать на данный момент можно только файлы с расширением `.js`(`.mjs`) и `.json`.

И если с первыми двумя всё понятно то вы можете задаться вопросом что за `JSON` такой. 

`JSON` - это лёгкий формат обмена данными созданный на основе JavaScript. 

Пример `JSON` кода:

```json
{
  "имя": "Глеб", 
  "Возраст": 13
}
```

Примечание: В `JSON` нет комментариев. 

[Подробнее здесь](https://github.com/TheFantomKiller420/CLEO-Redux-rus-local/blob/master/using-json.md). 

#### Свободный интерфейс

Методы конструируемых сущностей (таких как `Player`, `Car`, `Char` – любые сущности, созданные с помощью метода конструктора) поддерживают цепочку (также известную как Fluent Interface). Это позволяет писать такой код:

```js
var p = new Player(0);

p.giveWeapon(2, 100)
  .setHealth(5)
  .setCurrentWeapon(2)
  .getChar()
  .setCoordinates(1144, -600, 14)
  .setBleeding(true);
```

Посмотреть демонстрацию: https://www.youtube.com/watch?v=LLgJ0fWbklg.

Методы деструктора прерывают цепочку. Например. учитывая код:

`Char.Create(0, 0, 0, 0, 0).setCoordinates(0, 0, 0).delete()`

Цепочка не может продолжаться после метода удаления, так как символ удаляется, а его дескриптор больше не действителен.

#### Примеры

Если вы должны были изменить время суток на 22:00, выполните следующую команду

```js
Clock.SetTimeOfDay(22, 0);
```
Это будет эквивалентно коду операции `00C0: set_time_of_day 22 0` или `SET_TIME_OF_DAY 22 0` в коде SCM.

Поменять деньги игрока

```js
var player = new Player(0);
player.addScore(1000);
```

`new Player(0);` эквивалентен получению индекса игрока с использованием глобальной переменной `$PLAYER_CHAR` в Sanny Builder.

Создадим динамический объект (например, автомобиль)

```js
// запрашиваем модель 
Streaming.RequestModel(101);

// подождём пока игра загрузит модель
while (!Streaming.HasModelLoaded(101)) {
  wait(250);
}

// создаём машину по координатам и сохранить её хэндл
var infernus = Car.Create(101, 1234.0, 567.0, -100.0);
```

Переменная `infernus` будет содержать хэндл только что созданного автомобиля. Его можно передать напрямую любой команде, получающей дескриптор `Car`.

```js
// создать маркер на созданном автомобиле и сохранить хэндл маркера
var blip = Blip.AddForCar(infernus);
```

Эту же переменную infernus можно использовать для вызова методов `Car`'а, например, чтобы взорвать машину.

```js
infernus.explode();
```

Это будет эквивалентно коду операции `020B: взорвать_автомобиль $infernus` или `EXPLODE_CAR $infernus` в сценариях SCM.

Также ознакомьтесь с [примерами скриптов](examples). 

### Импорт

Вы можете импортировать другие файлы сценариев в свой код, чтобы сделать код модульным и использовать общую логику. Среда выполнения поддерживает оператор import, как описано в https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import.

Чтобы избежать запуска включенных файлов `.js` как отдельных скриптов, либо поместите их в отдельную папку (например, `CLEO/includes/`), либо используйте расширение `.mjs`.

```js
import func from "./other"; // импортируем экспорт по умолчанию из other.js или other.mjs, расположенных в том же каталоге

import { PedType } from "./includes/types"; // импортируем именованный экспорт PedType из types.js или types.mjs, расположенных в каталоге CLEO/includes

import data from "./vehicles.json"; // импортируем cars.json как значение JavaScript (массив, объект).
```

В настоящее время поддерживается только импорт файлов `.js` (`.mjs`) и `.json`.

### Пользовательские привязки

Следующие переменные и функции доступны только в коде JavaScript.

- `GAME` - идентификатор текущей игры. Возможные значения: `gta3`, `vc`, `re3`, `reVC`, `sa`, `gta3_unreal`, `vc_unreal`, `sa_unreal`.

```js
if (GAME === "gta3") {
  showTextBox("This is GTA III");
}
if (GAME === "sa") {
  showTextBox("This is San Andreas");
}
if (GAME === "sa_unreal") {
  showTextBox("This is San Andreas: The Definitive Edition");
}
```

- `ONMISSION` - глобальный флаг, определяющий, находится ли игрок в данный момент на миссии.

```js
if (!ONMISSION) {
  showTextBox("Not on a mission. Setting ONMISSION to true");
  ONMISSION = true;
}
```

- `TIMERA` и `TIMERB` - два автоматически увеличивающихся таймера, полезных для измерения временных интервалов.

```js
while (true) {
  TIMERA = 0;
  // ждём 1000 мс
  while (TIMERA < 1000) {
    wait(0);
  }
  showTextBox("1 second passed");
}
```


CHANGELOG.md
### 0.9.3 - Apr 22, 2022

- CLEO можно [встроить](https://re.cleo.li/docs/en/embedding.html) и запускать на неизвестных хостах в режиме самообслуживания [см. демонстрацию](https://www.youtube.com/watch?v=rk2LvDt7UkI)
- новый установщик, который автоматически загружает дополнительные зависимости, такие как Ultimate ASI Loader и плагины (dylib, IniFiles или ImGuiRedux)
- поддержка организации скриптов и их зависимостей в подкаталогах внутри папки CLEO. Подробнее: https://re.cleo.li/docs/en/script-lifecycle.html#organizing-scripts
- автоматически загружается последний файл `enums.js` из библиотеки Sanny Builder вместе с определениями команд. Вы можете импортировать перечисления в JS с помощью `import * as enums from './.config/enums';` 
- операции доступа к памяти могут выполняться на неизвестном хосте (ранее они зависели от команды `op`, которая сама могла выполняться только в играх GTA)
- `Memory.CallFunctionReturnFloat` и `Memory.CallMethodReturnFloat` теперь доступны для 32-разрядных хостов. `CallFunctionReturnFloat` ранее был добавлен для 64-битных хостов.

**SDK И ПЛАГИНЫ**
- Метод SDK `ResolvePath` теперь разрешает пути, начинающиеся с `./` или `.\` относительно каталога сценария. Вы можете использовать их в таких командах, как `READ_INT_FROM_INI_FILE` или `LOAD_DYNAMIC_LIBRARY`
- новые методы SDK `GetHostName`, `SetHostName`, `RuntimeInit`, `RuntimeNextTick`. Версия SDK теперь 2.
- Плагин IniFiles обновлен до версии 1.2: увеличена максимальная длина пути к файлу INI.
- Плагин Dylib обновлен до версии 1.1: увеличена максимальная длина пути к файлу DLL.

**ОСНОВНЫЕ ИЗМЕНЕНИЯ**
- удалена ранее устаревшая команда `op`. Вместо этого используйте `native`.
- перепенная `GAME` переименована в `HOST` (`GAME` по-прежнему доступна для использования, но рекомендуется обновить старые скрипты)

| Игра                                | Файл                                                                                                 | Минимальная требуемая версия |
| ----------------------------------- | ---------------------------------------------------------------------------------------------------- | ------------------------------ |
| GTA III, re3                        | [gta3.json](https://github.com/sannybuilder/library/blob/master/gta3/gta3.json)                      | `0.218`                        |
| GTA VC, reVC                        | [vc.json](https://github.com/sannybuilder/library/blob/master/vc/vc.json)                            | `0.220`                        |
| GTA San Andreas (Classic) 1.0       | [sa.json](https://github.com/sannybuilder/library/blob/master/sa/sa.json)                            | `0.237`                        |
| GTA III: The Definitive Edition     | [gta3_unreal.json](https://github.com/sannybuilder/library/blob/master/gta3_unreal/gta3_unreal.json) | `0.213`                        |
| Vice City: The Definitive Edition   | [vc_unreal.json](https://github.com/sannybuilder/library/blob/master/vc_unreal/vc_unreal.json)       | `0.215`                        |
| San Andreas: The Definitive Edition | [sa_unreal.json](https://github.com/sannybuilder/library/blob/master/sa_unreal/sa_unreal.json)       | `0.221`                        |

### 0.9.2 - Mar 04, 2022 
  
- Добавлена поддержка для The Definitive Edition Название обновления 1.04 (GTA III DE 1.0.0.15284, VC DE 1.0.0.15399, SA DE 1.0.0.15483) 
- Исправлена проблема со строковыми аргументами в командах вызова памяти в GTA San Andreas (https://github.com/cleolibrary/CLEO-Redux/issues/36) 
- Исправлена проблема с неработающими сценариями, если путь к каталогу с игрой заключен в квадратные скобки `[`, `]` 
  
### 0.9.1 - Feb 22, 2022

- Добавлен [SDK для разработки пользовательских команд](using-sdk.md) с помощью C++ и Rust
- Добавлена поддержка ошибочных команд в JS (также известных как команды `IF и SET` в сценариях SCM), они возвращают `undefined` при сбое, например.  `DynamicLibrary.Load` или `Char.IsInAnySearchlight`)
- Два новых плагина, добавляющих команды для работы с DLL (`dylib.cleo`) и INI-файлами (`IniFiles.cleo`) во всех поддерживаемых играх
- Добавлена константа `__dirname` в сценарии JS, которая разрешается в каталог текущего файла
- Добавлена новая функция [native](readme.md#пользовательские-привязки), которая вызывает команду сценария по имени (аналогично `op`):

```ts
/** Выполняет именованную команду с заданными аргументами */
declare function native<T>(name: string, ...args: any[]): T;
```

```js
const lib = native("LOAD_DYNAMIC_LIBRARY", "test.dll");
```

- Исправлена проблема с округлением чисел с плавающей точкой в GTA 3
- Исправлена проблема с неработающим импортом в JS, когда папка CLEO находится в каталоге AppData
- Исправлена ошибка, из-за которой команда `showTextBox` в San Andreas отображала мусорный текст
- Исправлена ошибка преобразования, когда команде `showTextBox` дается целочисленный аргумент
- Исправлена проблема с разрешениями сценариев, которые не проверялись для сценариев JS
- Исправлена проблема, когда объект, возвращаемый в результате выполнения некоторых команд (`Object.GrabEntityOnRope()`, `Heli.GrabEntityOnWinch()` и т.п.), не имел соответствующих полей, завернутых в экземпляр класса

**ЭТАПЫ УСТАНОВКИ**

https://github.com/cleolibrary/CLEO-Redux/blob/master/README.md#installation

**ОСНОВНЫЕ ИЗМЕНЕНИЯ**

| Игра                                | Файл                                                                                                 | Минимальная требуемая версия |
| ----------------------------------- | ---------------------------------------------------------------------------------------------------- | ---------------------------- |
| GTA III, re3                        | [gta3.json](https://github.com/sannybuilder/library/blob/master/gta3/gta3.json)                      | `0.208`                      |

examples/Algorithms.js
/// <reference path=".config/sa.d.ts" />

function binarySearch(sortedArray, key){
    let start = 0;
    let end = sortedArray.length - 1;

    while (start <= end) {
        let middle = Math.floor((start + end) / 2);

        if (sortedArray[middle] === key) {
            return "Index: " + middle + "; content: " + sortedArray[middle];
        } else if (sortedArray[middle] < key) {
            start = middle + 1;
        } else {
            end = middle - 1;
        }
    }
    return -1;
}

function linearSearch(arr, key){
    for(let i = 0; i < arr.length; i++){
        if(arr[i] === key){
            return "Index: " + i + "; content: " + arr[i];
        }
    }
    return null;
}

function bubbleSort(arr) {
    for (let j = arr.length - 1; j > 0; j--) {
        for (let i = 0; i < j; i++) {
            if (arr[i] > arr[i + 1]) {
                let temp = arr[i];
                arr[i] = arr[i + 1];
                arr[i + 1] = temp;
            }
        }
    }
    return arr;
}

function gcd(a, b) {
	if (a === 0)
		return Math.abs(b);
	if (b === 0)
		return Math.abs(a);
	for ( ; ; )
		if ((a %= b) === 0)
			return Math.abs(b);
		else if ((b %= a) === 0)
			return Math.abs(a);
}

function GCDBinary(a, b) {
	var factor = 1;
	while (true) {
		if (a == b)
			if (a == 0)
                throw 'GCD(0, 0)'
			else
                return factor * a;
		if (a == 0)
            return factor * b;
		if (b == 0)
            return factor * a;
		if (a == 1 || b == 1)
            return factor;
		if (!(a & 1) && !(b & 1)){
			factor <<= 1;
			a >>= 1;
			b >>= 1;
		}
		else if (!(a & 1))
            a >>= 1;
		else if (!(b & 1))
            b >>= 1;
		else if (b > a)
           b = (b - a) >> 1;		
		else
            a = (a - b) >> 1;
	}
}

// showTextBox(binarySearch([1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100], 56));

// showTextBox(linearSearch(["big", 5, 50, "small", false, 15.7], false));

// showTextBox(bubbleSort([25000,6,45,5462,4745,247,3,4688,7,5994]));

// showTextBox(gcd(8, 4));

// showTextBox(GCDBinary(5, 10));